src/modules/navigator/mission_reverse.cpp

```diff
@@ -0,0 +1,399 @@
+
+	if (!(_mission->_offboard_mission.count == 0)) {
+		if (_navigator->get_vstatus()->is_rotary_wing &&
+		    _navigator->get_vstatus()->is_vtol &&
+		    !_navigator->get_land_detected()->landed) {
+			command_vtol_transition();
+
+		} else {
+			advance_mission();
+			set_mission_items();
+		}
+	}
+
+	mavlink_log_info(_navigator->get_mavlink_log_pub(), "Mission Reverse activated");
+}
+
+void
+MissionReverse::on_active()
+{
+	// don't do anything if the plane is landed
+	if (_navigator->get_land_detected()->landed) {
+		return;
+	}
+
+	if (mission_changed()) {
+		_mission->_current_offboard_mission_index = index_closest_mission_item();
+		set_mission_items();
+	}
+
+	if (_mission->_mission_type != Mission::MISSION_TYPE_NONE && _mission->is_mission_item_reached()) {
+		/* If we just completed a takeoff which was inserted before the right waypoint,
+		   there is no need to report that we reached it because we didn't. */
+		if (_mission->_work_item_type != Mission::WORK_ITEM_TYPE_TAKEOFF) {
+			_mission->set_mission_item_reached();
+		}
+
+		// autocontinue to the next waypoint
+		advance_mission();
+		set_mission_items();
+
+	} else if ((_mission->_mission_type != Mission::MISSION_TYPE_NONE)
+		   && _mission->_param_altmode.get() == Mission::MISSION_ALTMODE_FOH) {
+		_mission->altitude_sp_foh_update();
+
+	} else {
+		/* if waypoint position reached allow loiter on the setpoint */
+		if (_mission->_waypoint_position_reached && _mission->_mission_item.nav_cmd != NAV_CMD_IDLE) {
+			_navigator->set_can_loiter_at_sp(true);
+		}
+	}
+
+	/* check if a cruise speed change has been commanded */
+	if (_mission->_mission_type != Mission::MISSION_TYPE_NONE) {
+		_mission->cruising_speed_sp_update();
+	}
+
+	/* see if we need to update the current yaw heading */
+	if ((_mission->_param_yawmode.get() != Mission::MISSION_YAWMODE_NONE
+	     && _mission->_param_yawmode.get() < Mission::MISSION_YAWMODE_MAX
+	     && _mission->_mission_type != Mission::MISSION_TYPE_NONE)
+	    || _navigator->get_vstatus()->is_vtol) {
+
+		_mission->heading_sp_update();
+	}
+
+	/* check if landing needs to be aborted */
+	if ((_mission_item.nav_cmd == NAV_CMD_LAND)
+	    && (_navigator->abort_landing())) {
+
+		_mission->do_abort_landing();
+	}
+}
+
+void
+MissionReverse::advance_mission()
+{
+	dm_item_t dm_current = (dm_item_t)(_mission->_offboard_mission.dataman_id);
+
+	for (int32_t i = _mission->_current_offboard_mission_index - 1; i >= 0; i--) {
+		struct mission_item_s missionitem = {};
+		const ssize_t len = sizeof(missionitem);
+
+		if (dm_read(dm_current, i, &missionitem, len) != len) {
+			/* not supposed to happen unless the datamanager can't access the SD card, etc. */
+			break;
+		}
+
+		if (item_contains_position(missionitem)) {
+			_mission->_current_offboard_mission_index = i;
+			return;
+		}
+	}
+
+	// finished flying back the mission
+	_mission->_current_offboard_mission_index = -1;
+}
+
+void
+MissionReverse::set_mission_items()
```

Comment:
Could we directly call Mission::set_mission_items() instead if the next next position setpoint is first set up?